### PR TITLE
Only accessors

### PR DESCRIPTION
**Pull request by davidanthoff**
This should be the uncontroversial stuff. ~~I'll have one more look before it is ready for review.~~

Ready!

### CODE REVIEW

File: src/requests/textdocument.jl [hunk 1]
```diff
@@ -28,38 +29,38 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didOpen")},DidOpenT
...
+#         end
+#     end
+#     get_line_offsets(doc)
+#     parse_all(doc, server)
+# end
 
 JSONRPC.parse_params(::Type{Val{Symbol("textDocument/didClose")}}, params) = DidCloseTextDocumentParams(params)
 function process(r::JSONRPC.Request{Val{Symbol("textDocument/didClose")},DidCloseTextDocumentParams}, server)
     uri = r.params.textDocument.uri
-    !haskey(server.documents, URI2(uri)) && return
```
Comment (davidanthoff) [inline src/requests/textdocument.jl]: I think this just masked bugs, so I removed.

File: src/requests/workspace.jl [hunk 2]
```diff
@@ -2,17 +2,21 @@ JSONRPC.parse_params(::Type{Val{Symbol("workspace/didChangeWatchedFiles")}}, par
 function process(r::JSONRPC.Request{Val{Symbol("workspace/didChangeWatchedFiles")},DidChangeWatchedFilesParams}, server)
     for change in r.params.changes
         uri = change.uri
-        !haskey(server.documents, URI2(uri)) && continue
-        if change.type == FileChangeTypes["Created"] || (change.type == FileChangeTypes["Changed"] && !get_open_in_editor(server.documents[URI2(uri)]))
-            doc = server.documents[URI2(uri)]
+        # TODO DA This seems like a bug, in the case of a create we should
+        # read the file from disc and add it to documents.
+        !hasdocument(server, URI2(uri)) && continue
```
Comment (davidanthoff) [inline src/requests/workspace.jl]: We can fix this in a follow up PR.

File: src/requests/textdocument.jl [hunk 3]
```diff
@@ -28,38 +29,38 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didOpen")},DidOpenT
...
-    doc = server.documents[URI2(uri)]
+    doc = getdocument(server, URI2(uri))
     empty!(doc.diagnostics)
     publish_diagnostics(doc, server)
     if !is_workspace_file(doc)
-        delete!(server.documents, URI2(uri))
+        deletedocument!(server, URI2(uri))
```
Comment (ZacLN) [inline src/requests/textdocument.jl]: Note to self: should clear meta and re-run staticlint pass across project roots.

File: src/requests/misc.jl [hunk 4]
```diff
@@ -9,7 +9,7 @@ function process(r::JSONRPC.Request{Val{Symbol("julia/lint-package")},Nothing},
 
 JSONRPC.parse_params(::Type{Val{Symbol("julia/toggle-lint")}}, params) = TextDocumentIdentifier(params["textDocument"])
 function process(r::JSONRPC.Request{Val{Symbol("julia/toggle-lint")},TextDocumentIdentifier}, server)
-    doc = server.documents[URI2(r.uri)]
+    doc = getdocument(server, URI2(r.uri))
     doc._runlinter = !doc._runlinter
 end
```
Comment (ZacLN) [inline src/requests/misc.jl]: Note: Should remove from here and julia-vscode.

File: src/requests/features.jl [hunk 5]
```diff
@@ -157,7 +148,7 @@ end
 
 JSONRPC.parse_params(::Type{Val{Symbol("textDocument/documentLink")}}, params) = DocumentLinkParams(params) 
 function process(r::JSONRPC.Request{Val{Symbol("textDocument/documentLink")},DocumentLinkParams}, server)
-    if !haskey(server.documents, URI2(r.params.textDocument.uri))
+    if !hasdocument(server, URI2(r.params.textDocument.uri))
         error("Received 'textDocument/documentLink for non-existing document.")
     end
     links = Tuple{String,UnitRange{Int}}[]
```
Comment (ZacLN) [inline src/requests/features.jl]: Note: not currently supported in InitOptions so comment out/remove

File: src/requests/textdocument.jl [hunk 6]
```diff
@@ -332,14 +330,14 @@ end
...
-    doc = server.documents[uri]
+    doc = getdocument(server, uri)
     empty!(doc.diagnostics)
     publishDiagnosticsParams = PublishDiagnosticsParams(doc._uri, doc._version, Diagnostic[])
     JSONRPCEndpoints.send_notification(server.jr_endpoint, "textDocument/publishDiagnostics", publishDiagnosticsParams)
 end 
 
 function clear_diagnostics(server)
-    for (uri, doc) in server.documents
+    for uri in getdocuments_key(server)
```
Comment (ZacLN) [inline src/requests/textdocument.jl]: Note: this function should probably operate on Documents

**Comment by davidanthoff** [comment]
Ok, I merged, and we can then address the open issues with smaller, more targeted PRs.

File: src/requests/misc.jl (hunk 4 above)
Comment (davidanthoff) [reply src/requests/misc.jl]: https://github.com/julia-vscode/LanguageServer.jl/pull/541. As far as I can tell there was no corresponding code in julia-vscode anymore, we must have removed it there already.

File: src/requests/features.jl (hunk 5 above)
Comment (davidanthoff) [reply src/requests/features.jl]: https://github.com/julia-vscode/LanguageServer.jl/pull/542

**Pull request merged**